Circular kernel

ferrine opened the pull request:
In this PR Circular kernel in introduced to work on exotic domains, e.g. Unit Circle

+ [x] what are the (breaking) changes that this PR makes?
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] consider adding/updating relevant example notebooks
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

`pymc3.gp.cov.Circular` is a new class to work with. The example can be found in the supplementary notebook.

ferrine committed:
Adding Cirdular kernel

ferrine committed:
adding an example notebook

ferrine committed:
ad docs, append to release notes

AlexAndorra commented on RELEASE-NOTES.md:
```diff
@@ -11,7 +11,7 @@
 ### New features
 - `sample_posterior_predictive_w` can now feed on `xarray.Dataset` - e.g. from `InferenceData.posterior`. (see [#4042](https://github.com/pymc-devs/pymc3/pull/4042))
-
+- added `pymc3.gp.cov.Circular` see [#4082](https://github.com/pymc-devs/pymc3/pull/4082), kernel for circular domains (e.g. unit circle)
```
```suggestion
- Added `pymc3.gp.cov.Circular` kernel for Gaussian Processes on circular domains, e.g. the unit circle (see [#4082](https://github.com/pymc-devs/pymc3/pull/4082)).
```

AlexAndorra reviewed (changes requested):
Thanks a lot @ferrine , super exciting to get this into PyMC3 🎉 
This looks good to me, I just left a few comments below, and going to review the associated NB now 😉

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<p>Can you style the NB according to <a href="https://github.com/pymc-devs/pymc3/wiki/PyMC's-Jupyter-Notebook-Style" target="_blank">the style guide</a> please?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

ferrine replied:
sure, I'm on it

ferrine replied:
done

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li>"Circular domains are <strong>a</strong> challenge for Gaussian Processes. Periodic patterns are assumed, but they are hard to capture with primitives <strong>and combining them is computationally expensive. An a</strong>lternative approach is required."</li><li>Also, could be interesting to add a few words about <em>why</em> circular domains are a challenge for GPs?</li><li>"<strong>the</strong> Weinland function is used to solve the problem and ensure<s>s</s> positive definite kernel<strong>s</strong> on the circular domain"</li><li>I suppose <code>t</code> is time? Would be interesting to add this precision, as this would also enlighten us as to the concrete meaning of <code>c</code>. Also interesting to explain rapidly what <code>tau</code> represents</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='2'/>

AlexAndorra replied:
<ul><li>I don't understand this sentence: "For circular domain [0, 2π), how to model correlation between π−ε and ε, real distance is 2ε, but computes differently with simple math"</li><li>"For correctly computed distances we need to <s>ferify</s> <strong>verify the</strong> kernel we obtain is positive definite"</li></ul>

ferrine replied:
<p>updated</p>

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<p>Maybe explain quickly what geodesic means?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

ferrine replied:
done

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<p>Why is the second term different here than in the latex formula above?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='4'/>

AlexAndorra replied:
<p>Get rid of matplotlib object printing by adding a ";" at the end of the plot: <code>plt.legend();</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<p>Maybe add a short doc string to explain where this formula comes from and what it does?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='5'/>

ferrine replied:
it's copy paste from SO)))

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li>"<strong>As we see, the higher </strong><code><strong>tau</strong></code><strong> is, the less correlated the samples</strong>"</li><li>"<strong>Also, we can extrapolate, getting cycles over and over:</strong>"</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='8'/>

ferrine replied:
refactored

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li>I don't think this plot is understable at first sight (I had to put a lot of thought into it to get what it means). Some axis labels and / or text explanation could help?</li><li>Also you can get rid of matplotlib object printing by adding a ";" at the end of the plot</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='9'/>

AlexAndorra replied:
<p>Looks nice now, thanks! As a final touch, do you think you can add axis labels, to speed-up understanding by readers?</p>

ferrine replied:
<p>added labels, they are explainable now</p>

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li>Can you put the observed data and the <code>angles</code> object into context? This would help readers understand in which cases a circular kernel would be useful. I guess this would also help understand why you shift some variables by 4 in the model below</li><li>Would be nice to explain the conceptual difference btw <code>y_sampled</code> and <code>y_pred</code></li><li>Also, any reason you're not doing the full Bayesian sampling?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

ferrine replied:
<p>full bayes is too slow</p>

ferrine replied:
totally rewritten

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li><code>vars</code>  is deprecated, better to use <code>var_names</code>  instead</li><li>maybe let's not call the object <code>trace</code>,  as this is usually reserved to posterior samples, not posterior predictive -- could be confusing to readers</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

ferrine replied:
I've used varnames, but missed the `trace` suggestion

AlexAndorra commented on docs/source/notebooks/GP-Circular.ipynb:
```diff
@@ -0,0 +1,329 @@
+{
```
<ul><li>Why are you taking the square root of the variance in the second <code>fill_between</code> but not in the first one?</li><li>You can rid of matplotlib object printing by adding a ";" at the end of the plot</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

ferrine replied:
done

twiecki commented:
Also should be added to the GP kernel NB.

ferrine committed:
update circular example

ferrine committed:
Update RELEASE-NOTES.md

Co-authored-by: Alexandre ANDORRA <andorra.alexandre@gmail.com>

ferrine committed:
review nb

ferrine committed:
Merge branch 'circular-kernel' of github.com:pymc-devs/pymc3 into circular-kernel

ferrine committed:
fix typos

ferrine committed:
circular update only

ferrine committed:
update kernels and covs nb

ferrine committed:
fix typos

ferrine committed:
bound->period for consistency with Periodic

ferrine committed:
update plots, notebooks

ferrine committed:
fix typo in doc

ferrine committed:
fix even more typos

ferrine committed:
Merge branch 'master' into circular-kernel

ferrine commented:
I've reviewed the circular kernel once more. In the paper parameterization, the lengthscale parameter appeared to cancel out in computation. I've updated the notebooks and added a comment to the Circular-GP notebook about that

ferrine committed:
follow pre-commit advice

ferrine committed:
typo fix

AlexAndorra reviewed (commented):
Thanks @ferrine, really like the new version with the comparisons of different kernels, the take-aways and the explanation 👏 
If you don't mind, I think there are still some typos / stuff to explicit, that I flagged above 😉 Tell me if anything is unclear. I think it'll be good to merge after that!

ferrine commented:
Sure

ferrine commented:
<p>y_sampled is used for <code>trace = pm.sample_posterior_predictive([mp], var_names=["y"], samples=100)</code>  but the actual variable is recorded my the model context</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ferrine commented:
<blockquote>why is the GP's mean set to 4?</blockquote><p>It looks better on the plots</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/4082/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ferrine committed:
update notebooks

ferrine commented:
The morning following @AlexAndorra's comments is the best morning

ferrine committed:
nitpics

ferrine committed:
fix typo

AlexAndorra reviewed (approved):
Amazing, thanks a lot @ferrine 👏 And thanks for bearing with me and all my comments 😄
